File: android/app/src/main/java/com/xiangshang/youth/core/util/BusinessClock.kt
package com.xiangshang.youth.core.util

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Canonical business clock for records shared with the school backend.
 *
 * Dates such as check-ins, assessment receipts and growth windows are a
 * school-day concept, not the device's current timezone. Keep the formatter
 * short-lived because SimpleDateFormat is mutable and not thread-safe.
 */
object BusinessClock {
    const val timeZoneId = "Asia/Shanghai"
    val timeZone: TimeZone = TimeZone.getTimeZone(timeZoneId)

    fun format(pattern: String, date: Date = Date()): String =
        SimpleDateFormat(pattern, Locale.US).apply {
            isLenient = false
            timeZone = BusinessClock.timeZone
        }.format(date)

    fun day(date: Date = Date()): String = format("yyyy-MM-dd", date)
}

File: ios/XiangshangYouth/Core/Extensions/BusinessClock.swift
import Foundation

/// Canonical school-day clock shared by local records and growth calendars.
/// Device timezone must not move a check-in or follow-up across a school day.
enum BusinessClock {
    static let timeZone = TimeZone(identifier: "Asia/Shanghai")!

    static var calendar: Calendar {
        var value = Calendar(identifier: .gregorian)
        value.locale = Locale(identifier: "en_US_POSIX")
        value.timeZone = timeZone
        return value
    }

    static func startOfDay(_ date: Date = .now) -> Date { calendar.startOfDay(for: date) }

    static func addingDays(_ days: Int, to date: Date = .now) -> Date {
        calendar.date(byAdding: .day, value: days, to: date) ?? date
    }

    static func string(_ date: Date = .now, format: String) -> String {
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = timeZone
        formatter.dateFormat = format
        formatter.isLenient = false
        return formatter.string(from: date)
    }

    static func day(_ date: Date = .now) -> String { string(date, format: "yyyy-MM-dd") }
}
